### DISCUSSION TITLE
Retrieve local files

### DISCUSSION THREAD

**Discussion by sitadrost** [abraunegg/onedrive#1359] [Q&A] [answered]
I just started using OneDrive to back up my data, and initially everything seemed to work fine. However, when I did a synchronize just now, OLDER versions of my files were downloaded from OneDrive, and my latest work was overwritten!

Clearly there's some setting that I've got wrong, so now my questions are:

1. Most importantly: is there any way that I can retrieve my latest work?
2. Can the client be configured such that it prompts me before overwriting newer files with older ones?

Many thanks!
Sita

**Comment by abraunegg** [comment, marked as answer]
@sitadrost 
> However, when I did a synchronize just now, OLDER versions of my files were downloaded from OneDrive, and my latest work was overwritten!

This is *not* possible as the client, if a file exists locally, the existing file is backed up by renaming it to avoid data loss. The biggest omission here from your discussion is:
* How did you install the client
* What client version are you using

> Most importantly: is there any way that I can retrieve my latest work?

Not from your local filesystem if you do not have local backups or snapshots. If you want to use snapshots, you can change your local filesystem to use ZFS, then utilise snapshots on your data which would prevent this sort of data loss in the future.

> Can the client be configured such that it prompts me before overwriting newer files with older ones?

As per above, the client *never* overwrites an existing file. It renames the existing file if an existing file is found to be in conflict.

Please can you check what client version you have installed as what you are claiming is 100% odd.

**Comment by sitadrost** [comment]
Sorry, you're right, I should have given more details. 
Version: onedrive v2.4.10-16-g6bec0dd
Installed on Ubuntu 20.04 using the instructions [here](https://gist.github.com/starlinq/0f98c6d9339497bb8ac42d67f66f60eb), haven't changed anything in the configuration yet:
```
Configuration file successfully loaded
onedrive version                       = v2.4.10-16-g6bec0dd
Config path                            = /home/sita/.config/onedrive
Config file found in config path       = true
Config option 'check_nosync'           = false
Config option 'sync_dir'               = /home/sita/OneDrive
Config option 'skip_dir'               = 
Config option 'skip_file'              = ~*|.~*|*.tmp
Config option 'skip_dotfiles'          = false
Config option 'skip_symlinks'          = false
Config option 'monitor_interval'       = 300
Config option 'min_notify_changes'     = 5
Config option 'log_dir'                = /var/log/onedrive/
Config option 'classify_as_big_delete' = 1000
Config option 'upload_only'            = false
Config option 'no_remote_delete'       = false
Config option 'remove_source_files'    = false
Config option 'sync_root_files'        = false
Selective sync 'sync_list' configured  = false
Business Shared Folders configured     = false

```
My local OneDrive folder is in my home directory, in this folder I created symlinks to the data I want to back up.

What you write about the client re-naming a file if it exists locally sounds promising, it might save me hours of re-writing the code I lost. I haven't found any re-named files yet though; where would these normally be stored?

Thanks a lot so far!

**Reply by abraunegg** [reply]
@sitadrost 
> Installed on Ubuntu 20.04 using the instructions here, haven't changed anything in the configuration yet

Dont follow those instructions - that gist is incorrect. You need to follow these instructions: https://github.com/abraunegg/onedrive/blob/master/docs/INSTALL.md

This being said, you are running at least close to the latest client - unlike most folk who use the Ubuntu packages which install old and unsupported versions.

> What you write about the client re-naming a file if it exists locally sounds promising, it might save me hours of re-writing the code I lost. I haven't found any re-named files yet though; where would these normally be stored?

If the file exists at the target location, it will be renamed to include the hostname. For example, if I am downloading a file called 'data.txt' to my ~/OneDrive folder, and *if* that file already exists and the client cache DB does not know about it, it, it will be renamed:
```
data.txt -> data-<hostname>.txt
```
So then in that directory, there would now be two files:
```
data.txt  <- as downloaded
data-<hostname>.txt <- original local file
```

**Reply by sitadrost** [reply]
@abraunegg Thanks for getting back so fast. Unfortunately I don't have any such files. Could my problem be in using a symlink, i.e. that the timestamp of the symlink is used instead of the actual file's timestamp?

**Reply by abraunegg** [reply]
@sitadrost 
> Could my problem be in using a symlink, i.e. that the timestamp of the symlink is used instead of the actual file's timestamp?

Doubtful - as link, or otherwise, is *still* a valid file.

What I would do is start from scratch and re-create your scenario. Then, run the client in `--synchronize --verbose --dry-run` mode and see what is going to happen. Additionally do not use `--resync` unless the application tells you to.

If you are then comfortable with what dry-run is doing, remove it and see what occurs, but to ensure you have the debug logs, run the client in `--synchronize --verbose --verbose` mode.

If you see something unexpected or your data / local file is being overwritten - you have the debug log to show you what is occurring.

Secondly, just confirm you *only* have one client installed.

When issues like this in the past have popped up - users have followed a number of 'websites' and installed not only this client, but the skilion one as well, and it is running in the background, causing deletion issues .... just make sure you are not falling into that trap as well.

**Reply by sitadrost** [reply]
@abraunegg Right, it looks like I'd better start from scratch. Little hope of restoring my code, so I'd better put some extra effort in preventing this from happening again.

I've no clue what skilion is, but as far as I can tell, it's not installed/running (according to `which` and `pgrep`). Where should I be looking for this?

**Reply by abraunegg** [reply]
@sitadrost 
> I've no clue what skilion is, but as far as I can tell, it's not installed/running

Skilion OneDrive Client: https://github.com/skilion/onedrive

If finding any binary called 'onedrive' only yields one binary and it is the one you installed, and it is v2.4.10 or greater - then all good. In the past:
* Folk have installed from package a very old client, usually because they are using Ubuntu and just ran `apt install onedrive` (which is 100% not supported as it installs an old client)
* To fix issues, installed they then installed 'skilion' client manually
* Found further issues with that client, then installed this client - either manually or via the Ubuntu PPA

This ended up with 3 x different binaries - and, based on search path precedence, would use one of the older clients, despite a newer one being installed.

> Right, it looks like I'd better start from scratch.

Yes, probably a good idea, but make sure you test with `--dry-run` everything *before* you do things so you can be sure of what is going to happen.

Only enable systemd services *once* everything is working manually (`--synchronize`)as per your expectation. Only use `--resync` if the application advises you to.

**Reply by sitadrost** [reply]
@abraunegg Good, that doesn't seem to be the case; I've only one onedrive client. 

Yes, I'll just start over from scratch. Even though I just did a dry-run with my current install and everything looked fine, I'd still prefer to have the client installed the proper way, to prevent the update issues you pointed out earlier.

Thanks for all your help!

**Reply by asterixix** [reply]
> @sitadrost
> 
> > Installed on Ubuntu 20.04 using the instructions here, haven't changed anything in the configuration yet
> 
> Dont follow those instructions - that gist is incorrect. You need to follow these instructions: https://github.com/abraunegg/onedrive/blob/master/docs/INSTALL.md
> 
> This being said, you are running at least close to the latest client - unlike most folk who use the Ubuntu packages which install old and unsupported versions.
> 
> > What you write about the client re-naming a file if it exists locally sounds promising, it might save me hours of re-writing the code I lost. I haven't found any re-named files yet though; where would these normally be stored?
> 
> If the file exists at the target location, it will be renamed to include the hostname. For example, if I am downloading a file called 'data.txt' to my ~/OneDrive folder, and _if_ that file already exists and the client cache DB does not know about it, it, it will be renamed:
> 
> ```
> data.txt -> data-<hostname>.txt
> ```
> 
> So then in that directory, there would now be two files:
> 
> ```
> data.txt  <- as downloaded
> data-<hostname>.txt <- original local file
> ```

Hi, 
I just ask because I'm totally dump about scripting on bash... How can I resolve that issue with duplicated files? 
I have sync that files with online onedrive and I want to remove that copies.

**Reply by abraunegg** [reply]
> > @sitadrost
> > > Installed on Ubuntu 20.04 using the instructions here, haven't changed anything in the configuration yet
> > 
> > 
> > Dont follow those instructions - that gist is incorrect. You need to follow these instructions: https://github.com/abraunegg/onedrive/blob/master/docs/INSTALL.md
> > This being said, you are running at least close to the latest client - unlike most folk who use the Ubuntu packages which install old and unsupported versions.
> > > What you write about the client re-naming a file if it exists locally sounds promising, it might save me hours of re-writing the code I lost. I haven't found any re-named files yet though; where would these normally be stored?
> > 
> > 
> > If the file exists at the target location, it will be renamed to include the hostname. For example, if I am downloading a file called 'data.txt' to my ~/OneDrive folder, and _if_ that file already exists and the client cache DB does not know about it, it, it will be renamed:
> > ```
> > data.txt -> data-<hostname>.txt
> > ```
> > 
> > 
> > So then in that directory, there would now be two files:
> > ```
> > data.txt  <- as downloaded
> > data-<hostname>.txt <- original local file
> > ```
> 
> Hi,
> I just ask because I'm totally dump about scripting on bash... How can I resolve that issue with duplicated files?
> I have sync that files with online onedrive and I want to remove that copies.

Use your favorite search engine with 'linux remove duplicate files' .........

**Reply by sitadrost** [reply]
Good, I finally managed to rewrite most of the code I lost in this sync that went the wrong way, so now I have time to actually start from scratch. The instructions in INSTALL.md tell me to run
`sudo make uninstall`
and then to either delete the entire application state,
`rm -rf ~/.config/onedrive`
or to just delete the application key ("If you want to just delete the application key but keep the items database"),
`rm -f ~/.config/onedrive/refresh_token`

I intend to re-install the onedrive client after this uninstall, this time using the correct instructions. Would only deleting the application key be the correct/best option in that case?

**Reply by abraunegg** [reply]
@sitadrost 
In your case, just perform `sudo make uninstall` - no need to remove anything else

**Reply by sitadrost** [reply]
Perfect, thanks!
